lua/apisix/plugins/tcp-logger.lua

```diff
@@ -48,51 +53,92 @@ function _M.check_schema(conf)
     end
 
-    sock:settimeout(conf.timeout)
+    sock:settimeout(conf.timeout * 1000)
 
     local ok, err = sock:connect(conf.host, conf.port)
     if not ok then
-        core.log.error("failed to connect to TCP server: host[",
-                conf.host, "] port[", conf.port, "] ", err)
-        return
+        return nil, "failed to connect to TCP server: host[" ..
+            conf.host .. "] port[" .. tostring(conf.port) .. "] err: " .. err
     end
 
     if conf.tls then
         ok, err = sock:sslhandshake(true, conf.tls_options, false)
         if not ok then
-            core.log.error("failed to to perform TLS handshake to TCP server: host[",
-                    conf.host, "] port[", conf.port, "] ", err)
-            return
+            return nil, "failed to to perform TLS handshake to TCP server: host[" ..
+                conf.host .. "] port[" .. tostring(conf.port) .. "] err: " .. err
         end
     end
 
     ok, err = sock:send(log_message)
     if not ok then
-        core.log.error("failed to send data to TCP server: host[",
-                conf.host, "] port[", conf.port, "] ", err)
+        res = false
+        err_msg = "failed to send data to TCP server: host[" ..
+            conf.host .. "] port[" .. tostring(conf.port) .. "] err: " .. err
     end
 
     ok, err = sock:close()
     if not ok then
         core.log.error("failed to close the TCP connection, host[",
-                conf.host, "] port[", conf.port, "] ", err)
+            conf.host, "] port[", conf.port, "] ", err)
     end
+
+    return res, err_msg
 end
 
 
 function _M.log(conf)
-    return timer_at(0, log, conf, core.json.encode(log_util.get_full_log(ngx)))
+    local entry = log_util.get_full_log(ngx)
+
+    if not entry.route_id then
+        core.log.error("failed to obtain the route id for udp logger")
+        return
+    end
+
+    local log_buffer = buffers[entry.route_id]
+
+    -- If a logger is not present for the route, create one
+    if not log_buffer then
+        -- Generate a function to be executed by the batch processor
+        local func = function(entries, batch_max_size)
+            local data
+            if batch_max_size == 1 then
+                data = core.json.encode(entries[1]) -- encode as single {}
+            else
+                data = core.json.encode(entries) -- encode as array [{}]
+            end
+            return send_tcp_data(conf, data)
+        end
+
+        local config = {
+            name = conf.name,
+            retry_delay = conf.retry_delay,
+            batch_max_size = conf.batch_max_size,
+            max_retry_count = conf.max_retry_count,
+            buffer_duration = conf.buffer_duration,
+            inactive_timeout = conf.inactive_timeout,
+        }
+
+        local err
+        log_buffer, err = batch_processor:new(func, config)
+
+        if not log_buffer then
+            core.log.err("error when creating the batch processor: " .. err)
```

Comment:
we should avoid generating new object.

`core.log.error("error when creating the batch processor: ", err)`

Reply:
`core.log.err` is wrong, it should be `core.log.error`